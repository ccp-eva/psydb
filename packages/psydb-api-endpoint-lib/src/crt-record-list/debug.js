'use strict';
var debug = (scope) => (
    require('../debug')(`crt-record-list:${scope}`)
);

module.exports = debug;
