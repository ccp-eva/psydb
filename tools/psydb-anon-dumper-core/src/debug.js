'use strict';
var Debug = require('debug');
module.exports = (suffix) => (
    Debug(`psydb-db-anon-dumper:${suffix}`)
)
