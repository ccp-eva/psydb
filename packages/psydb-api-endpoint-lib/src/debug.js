'use strict';
var debug = (scope) => (
    require('debug')(`psydb:api-endpoint-lib:${scope}`)
);

module.exports = debug;
