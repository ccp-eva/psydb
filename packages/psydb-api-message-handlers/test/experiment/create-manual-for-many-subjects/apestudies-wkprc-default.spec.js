'use strict';
require('debug').enable('psydb*');
var {
    expect, inline, jsonify, ejson, omit,
    ObjectId, RootHandler
} = require('./utils');

describe(inline`
    experiment/create-manual-for-many-subjects apestudies-wkprc-default
`, function () {
    var db;
    beforeEach(async function () {
        await this.restore('2022-12-10__1635-wkprc');
        db = this.getDbHandle();
    });

    it('can create experiments', async function () {
        var [ sendMessage ] = this.createMessenger({ RootHandler });

        var studyId = await this.getId('study', {
            shorthand: 'WKPRC TEST A'
        });
        
        var locationId = await this.getId('location', {
            name: 'Schimpansen Aussengehege'
        });
        
        var subjectStarId = await this.getId('subject', {
            name: 'Star'
        }, { subChannels: true });

        var subjectSarahId = await this.getId('subject', {
            name: 'Sarah'
        }, { subChannels: true });
        
        var labOperatorId = await this.getId('personnel', {
            firstname: 'bob'
        }, { subChannels: true });
        
        var studyTopicIds = [
            await this.getId('studyTopic', {
                name: 'Collaboration'
            }),
            await this.getId('studyTopic', {
                name: 'Self-Awareness'
            })
        ];
        
        var koaContext = await sendMessage({
            type: 'experiment/create-manual-for-many-subjects',
            timezone: 'Europe/Berlin',
            payload: jsonify({
                labMethod: 'apestudies-wkprc-default',
                studyId,
                locationId,
                subjectData: [
                    {
                        subjectId: subjectStarId,
                        status: 'participated',
                        timestamp: new Date('2022-12-09T12:30:00Z'),
                        excludeFromMoreExperimentsInStudy: false,
                    },
                    {
                        subjectId: subjectSarahId,
                        status: 'participated',
                        timestamp: new Date('2022-12-10T14:45:00Z'),
                        excludeFromMoreExperimentsInStudy: false,
                    }
                ],
                labOperatorIds: [ labOperatorId ],
                studyTopicIds,
                experimentName: 'the exp name'
            })
        });

        var { body } = koaContext.response;
        var [ 
            starExperimentUpdate,
            sarahExperimentUpdate,
            starSubjectUpdate,
            sarahSubjectUpdate,
        ] = body.data;

        var starExperiment = await this.getRecord('experiment', {
            _id: starExperimentUpdate.channelId
        });
        //console.dir(ejson(starExperiment), { depth: null });
        
        expect(ejson(starExperiment)).to.containSubset({
            type: 'manual',
            realType: 'apestudies-wkprc-default'
        });

        expect(ejson(omit({
            from: starExperiment.state,
            paths: [ 'seriesId' ]
        }))).toMatchSnapshot();

        expect(starExperiment.state.seriesId).to.be.ok;

        /////////////////////////
        var sarahExperiment = await this.getRecord('experiment', {
            _id: sarahExperimentUpdate.channelId
        });
        //console.dir(ejson(sarahExperiment), { depth: null });
        
        expect(ejson(sarahExperiment)).to.containSubset({
            type: 'manual',
            realType: 'apestudies-wkprc-default'
        });

        expect(ejson(omit({
            from: sarahExperiment.state,
            paths: [ 'seriesId' ]
        }))).toMatchSnapshot();

        expect(sarahExperiment.state.seriesId).to.be.ok;

        ///////////////////////////

        var starSubject = await this.getRecord('subject', {
            _id: starSubjectUpdate.channelId
        });
        //console.dir(ejson(starSubject), { depth: null });
        expect(starSubject.type).to.equal('schimpansen');
        
        // NOTE: make sure nothing else got messed up
        expect(ejson(starSubject.gdpr.state)).toMatchSnapshot();
        expect(ejson(omit({
            from: starSubject.scientific.state,
            paths: [ 'internals.participatedInStudies' ]
        }))).toMatchSnapshot();

        var starParticipations = (
            starSubject.scientific.state.internals.participatedInStudies
        );
        //console.dir(ejson(starParticipations), { depth: null })
        // NOTE: its stars second experiment since whe already had one
        // on this fixture
        expect(String(starParticipations[1].experimentId))
            .to.equal(String(starExperiment._id));
        expect(ejson(omit({
            fromItems: starParticipations,
            paths: [ '_id', 'experimentId' ]
        }))).toMatchSnapshot()

        //////////////////

        var sarahSubject = await this.getRecord('subject', {
            _id: sarahSubjectUpdate.channelId
        });
        //console.dir(ejson(sarahSubject), { depth: null });
        expect(sarahSubject.type).to.equal('schimpansen');
        
        // NOTE: make sure nothing else got messed up
        expect(ejson(sarahSubject.gdpr.state)).toMatchSnapshot();
        expect(ejson(omit({
            from: sarahSubject.scientific.state,
            paths: [ 'internals.participatedInStudies' ]
        }))).toMatchSnapshot();

        var sarahParticipations = (
            sarahSubject.scientific.state.internals.participatedInStudies
        );
        //console.dir(ejson(sarahParticipations), { depth: null })
        // NOTE: its sarahs second experiment since whe already had one
        // on this fixture
        expect(String(sarahParticipations[1].experimentId))
            .to.equal(String(sarahExperiment._id));
        expect(ejson(omit({
            fromItems: sarahParticipations,
            paths: [ '_id', 'experimentId' ]
        }))).toMatchSnapshot()
    })    
})
