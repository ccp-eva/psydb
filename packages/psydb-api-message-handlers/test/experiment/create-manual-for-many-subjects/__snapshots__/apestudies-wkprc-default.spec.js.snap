// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`experiment/create-manual-for-many-subjects apestudies-wkprc-default can create experiments 1`] = `
Object {
  "color": "#000000",
  "experimentName": "the exp name",
  "experimentOperatorIds": Array [
    Object {
      "$oid": "6333f5fc1cc4009a891643a7",
    },
  ],
  "experimentOperatorTeamId": null,
  "generalComment": "",
  "interval": Object {
    "end": Object {
      "$date": "2022-12-09T12:30:00.000Z",
    },
    "start": Object {
      "$date": "2022-12-09T12:30:00.000Z",
    },
  },
  "isCanceled": false,
  "isPostprocessed": true,
  "locationId": Object {
    "$oid": "6333f8f41cc4009a891643d7",
  },
  "locationRecordType": "wkprc_location",
  "selectedSubjectGroupIds": Array [],
  "selectedSubjectIds": Array [
    Object {
      "$oid": "6333f8441cc4009a891643c7",
    },
  ],
  "studyId": Object {
    "$oid": "6333f8871cc4009a891643cb",
  },
  "studyRecordType": "wkprc_studie",
  "studyTopicIds": Array [
    Object {
      "$oid": "6333f6f71cc4009a891643ad",
    },
    Object {
      "$oid": "6333f7011cc4009a891643b0",
    },
  ],
  "subjectData": Array [
    Object {
      "comment": "",
      "excludeFromMoreExperimentsInStudy": false,
      "invitationStatus": "scheduled",
      "participationStatus": "participated",
      "subjectId": Object {
        "$oid": "6333f8441cc4009a891643c7",
      },
      "subjectType": "schimpansen",
    },
  ],
}
`;

exports[`experiment/create-manual-for-many-subjects apestudies-wkprc-default can create experiments 2`] = `
Object {
  "color": "#000000",
  "experimentName": "the exp name",
  "experimentOperatorIds": Array [
    Object {
      "$oid": "6333f5fc1cc4009a891643a7",
    },
  ],
  "experimentOperatorTeamId": null,
  "generalComment": "",
  "interval": Object {
    "end": Object {
      "$date": "2022-12-10T14:45:00.000Z",
    },
    "start": Object {
      "$date": "2022-12-10T14:45:00.000Z",
    },
  },
  "isCanceled": false,
  "isPostprocessed": true,
  "locationId": Object {
    "$oid": "6333f8f41cc4009a891643d7",
  },
  "locationRecordType": "wkprc_location",
  "selectedSubjectGroupIds": Array [],
  "selectedSubjectIds": Array [
    Object {
      "$oid": "6333f8371cc4009a891643c3",
    },
  ],
  "studyId": Object {
    "$oid": "6333f8871cc4009a891643cb",
  },
  "studyRecordType": "wkprc_studie",
  "studyTopicIds": Array [
    Object {
      "$oid": "6333f6f71cc4009a891643ad",
    },
    Object {
      "$oid": "6333f7011cc4009a891643b0",
    },
  ],
  "subjectData": Array [
    Object {
      "comment": "",
      "excludeFromMoreExperimentsInStudy": false,
      "invitationStatus": "scheduled",
      "participationStatus": "participated",
      "subjectId": Object {
        "$oid": "6333f8371cc4009a891643c3",
      },
      "subjectType": "schimpansen",
    },
  ],
}
`;

exports[`experiment/create-manual-for-many-subjects apestudies-wkprc-default can create experiments 3`] = `
Object {
  "custom": Object {},
}
`;

exports[`experiment/create-manual-for-many-subjects apestudies-wkprc-default can create experiments 4`] = `
Object {
  "comment": "",
  "custom": Object {
    "name": "Star",
  },
  "internals": Object {
    "invitedForExperiments": Array [],
    "isRemoved": false,
  },
  "systemPermissions": Object {
    "accessRightsByResearchGroup": Array [
      Object {
        "permission": "write",
        "researchGroupId": Object {
          "$oid": "6333f3591cc4009a89164347",
        },
      },
    ],
    "isHidden": false,
  },
  "testingPermissions": Array [],
}
`;

exports[`experiment/create-manual-for-many-subjects apestudies-wkprc-default can create experiments 5`] = `
Array [
  Object {
    "excludeFromMoreExperimentsInStudy": false,
    "experimentName": "Co-Op Experiment A",
    "experimentOperatorIds": Array [
      Object {
        "$oid": "6333f5d61cc4009a891643a1",
      },
    ],
    "locationId": Object {
      "$oid": "6333f8f41cc4009a891643d7",
    },
    "locationType": "wkprc_location",
    "principalInvestigatorId": Object {
      "$oid": "6333f5d61cc4009a891643a1",
    },
    "realType": "apestudies-wkprc-default",
    "status": "participated",
    "studyId": Object {
      "$oid": "6333f8871cc4009a891643cb",
    },
    "studyTopicIds": Array [
      Object {
        "$oid": "6333f6f71cc4009a891643ad",
      },
    ],
    "studyType": "wkprc_studie",
    "timestamp": Object {
      "$date": "2022-10-02T22:00:00.000Z",
    },
    "type": "manual",
  },
  Object {
    "excludeFromMoreExperimentsInStudy": false,
    "experimentName": "the exp name",
    "experimentOperatorIds": Array [
      Object {
        "$oid": "6333f5fc1cc4009a891643a7",
      },
    ],
    "locationId": Object {
      "$oid": "6333f8f41cc4009a891643d7",
    },
    "locationType": "wkprc_location",
    "realType": "apestudies-wkprc-default",
    "status": "participated",
    "studyId": Object {
      "$oid": "6333f8871cc4009a891643cb",
    },
    "studyTopicIds": Array [
      Object {
        "$oid": "6333f6f71cc4009a891643ad",
      },
      Object {
        "$oid": "6333f7011cc4009a891643b0",
      },
    ],
    "studyType": "wkprc_studie",
    "timestamp": Object {
      "$date": "2022-12-09T12:30:00.000Z",
    },
    "type": "manual",
  },
]
`;

exports[`experiment/create-manual-for-many-subjects apestudies-wkprc-default can create experiments 6`] = `
Object {
  "custom": Object {},
}
`;

exports[`experiment/create-manual-for-many-subjects apestudies-wkprc-default can create experiments 7`] = `
Object {
  "comment": "",
  "custom": Object {
    "name": "Sarah",
  },
  "internals": Object {
    "invitedForExperiments": Array [],
    "isRemoved": false,
  },
  "systemPermissions": Object {
    "accessRightsByResearchGroup": Array [
      Object {
        "permission": "write",
        "researchGroupId": Object {
          "$oid": "6333f3591cc4009a89164347",
        },
      },
    ],
    "isHidden": false,
  },
  "testingPermissions": Array [],
}
`;

exports[`experiment/create-manual-for-many-subjects apestudies-wkprc-default can create experiments 8`] = `
Array [
  Object {
    "excludeFromMoreExperimentsInStudy": false,
    "experimentName": "Co-Op Experiment A",
    "experimentOperatorIds": Array [
      Object {
        "$oid": "6333f5d61cc4009a891643a1",
      },
    ],
    "locationId": Object {
      "$oid": "6333f8f41cc4009a891643d7",
    },
    "locationType": "wkprc_location",
    "principalInvestigatorId": Object {
      "$oid": "6333f5d61cc4009a891643a1",
    },
    "realType": "apestudies-wkprc-default",
    "status": "participated",
    "studyId": Object {
      "$oid": "6333f8871cc4009a891643cb",
    },
    "studyTopicIds": Array [
      Object {
        "$oid": "6333f6f71cc4009a891643ad",
      },
    ],
    "studyType": "wkprc_studie",
    "timestamp": Object {
      "$date": "2022-10-02T22:00:00.000Z",
    },
    "type": "manual",
  },
  Object {
    "excludeFromMoreExperimentsInStudy": false,
    "experimentName": "the exp name",
    "experimentOperatorIds": Array [
      Object {
        "$oid": "6333f5fc1cc4009a891643a7",
      },
    ],
    "locationId": Object {
      "$oid": "6333f8f41cc4009a891643d7",
    },
    "locationType": "wkprc_location",
    "realType": "apestudies-wkprc-default",
    "status": "participated",
    "studyId": Object {
      "$oid": "6333f8871cc4009a891643cb",
    },
    "studyTopicIds": Array [
      Object {
        "$oid": "6333f6f71cc4009a891643ad",
      },
      Object {
        "$oid": "6333f7011cc4009a891643b0",
      },
    ],
    "studyType": "wkprc_studie",
    "timestamp": Object {
      "$date": "2022-12-10T14:45:00.000Z",
    },
    "type": "manual",
  },
]
`;
